crate_universe/src/utils/starlark/label.rs

```diff
@@ -193,66 +309,129 @@ mod test {
+    fn absolute_package() {
+        let label = Label::from_str("//package").unwrap();
+        assert_eq!(label.to_string(), "//package:package");
+        assert!(label.is_absolute());
+        assert_eq!(label.repository(), None);
+        assert_eq!(label.package(), Some("package"));
+        assert_eq!(label.target(), "package");
     }
 
     #[test]
-    fn root_label_no_repository() {
+    fn absolute_target() {
         let label = Label::from_str("//:target").unwrap();
         assert_eq!(label.to_string(), "//:target");
-        assert_eq!(label.repository, None);
-        assert_eq!(label.package, None);
-        assert_eq!(label.target, "target");
+        assert!(label.is_absolute());
+        assert_eq!(label.repository(), None);
+        assert_eq!(label.package(), None);
+        assert_eq!(label.target(), "target");
     }
 
     #[test]
-    fn root_label_no_slashes() {
-        let label = Label::from_str(":target").unwrap();
-        assert_eq!(label.to_string(), "//:target");
-        assert_eq!(label.repository, None);
-        assert_eq!(label.package, None);
-        assert_eq!(label.target, "target");
+    fn absolute_repository_package() {
+        let label = Label::from_str("@repo//package").unwrap();
+        assert_eq!(label.to_string(), "@repo//package:package");
+        assert!(label.is_absolute());
+        assert_eq!(label.repository(), Some("repo"));
+        assert_eq!(label.package(), Some("package"));
+        assert_eq!(label.target(), "package");
+    }
+
+    #[test]
+    fn absolute_repository_target() {
+        let label = Label::from_str("@repo//:target").unwrap();
+        assert_eq!(label.to_string(), "@repo//:target");
+        assert!(label.is_absolute());
+        assert_eq!(label.repository(), Some("repo"));
+        assert_eq!(label.package(), None);
+        assert_eq!(label.target(), "target");
+    }
+
+    #[test]
+    fn absolute_package_target() {
+        let label = Label::from_str("//package:target").unwrap();
+        assert_eq!(label.to_string(), "//package:target");
+        assert!(label.is_absolute());
+        assert_eq!(label.repository(), None);
+        assert_eq!(label.package(), Some("package"));
+        assert_eq!(label.target(), "target");
+    }
+
+    #[test]
+    fn invalid_empty() {
+        Label::from_str("").unwrap_err();
+        Label::from_str("//").unwrap_err();
+    }
+
+    #[test]
+    fn invalid_relative_repository_package_target() {
+        Label::from_str("@repo/package:target").unwrap_err();
+    }
+
+    #[test]
+    fn invalid_relative_repository_package() {
+        Label::from_str("@repo/package").unwrap_err();
+    }
+
+    #[test]
+    fn invalid_relative_repository_target() {
+        Label::from_str("@repo:target").unwrap_err();
+    }
+
+    #[test]
+    fn invalid_relative_package_target() {
+        Label::from_str("package:target").unwrap_err();
+    }
+
+    #[test]
+    fn full_label_bzlmod() {
```

Comment:
This is an interesting test-case... If you hadn't added the test case, I wouldn't mind (and happy to remove it), but I think that if we're going to care about this, we should preserve the `@@` at the start. So we may need to keep an `is_canonical` bool in the label, set it to whether the prefix of the label is `@@`, and prefix the stringification with `@@` (even if the repo is empty) if `is_canonical`.

(And if we're being pedantic, we also should technically preserve an `@` at the start of a label, so `@//foo:bar` should stringify as `@//foo:bar` not `//foo:bar`).

Maybe the real way of handling this is to make make `repository` be an
```rust
enum Repository {
    Canonical(String), // stringifies to `@@self.0` where `self.0` may be empty
    Explicit(String), // stringifies to `@self.0` where `self.0` may be empty
    Local, // stringifies to the empty string
}
```

Reply:
I like the `enum Repository` approach!  Also I didn't add this test, just moved it 😉